lib/extensions/storage_allowance/storage_allowance.js

```diff
@@ -0,0 +1,127 @@
+   * Make sure to call `next` when ready call deeper into middleware stack, before handling responses in your middleware.
+   */
+  handle = async () => {
+    const [method, uri] = getRouting(this.request, this._options);
+    const candidateResponse = this.response.getCandidate();
+    const salt = this._options.extensions.storage_allowance.salt;
+
+    if (method === 'POST' && (uri.pathname === 'oauth' || uri.pathname === 'pay2myapp')) {
+      await this._next();
+
+      const username = this.params.username;
+      const size = this.getSize(username);
+      let hasCapacity = true;
+      if (this._options.extensions.storage_allowance.max_bytes) {
+        hasCapacity = size < this._options.extensions.storage_allowance.max_bytes;
+      }
+
+      const location = candidateResponse.headers.Location;
+      if (!location) return;
+      const preamble = location.split('#')[0];
+      const coded = location.split('#')[1];
+      const decoded = qs.parse(coded);
+      const token = decoded.access_token;
+      let value = 'false';
+      if (hasCapacity) {
+        value = symmetricEncrypt(`${HAS_WRITE_CAPACITY_METADATA_VALUE_PREFIX}::${(new Date()).toUTCString()}`, salt);
+      }
+      const updatedToken = metadataUpsert(token, HAS_WRITE_CAPACITY_METADATA, value);
+      const newcoded = { ...decoded, access_token: updatedToken };
+      const encoded = qs.stringify(newcoded);
+      candidateResponse.headers.Location = `${preamble}#${encoded}`;
+
+      return;
+    }
+
+    let metadata = {};
+    if (this.request.headers.authorization) {
+      const token = decodeURIComponent(this.request.headers.authorization.split(/\s+/)[1]);
+      metadata = getMetadata(token);
+      const updatedToken = metadataDelete(token, HAS_WRITE_CAPACITY_METADATA);
+      this.request.headers.authorization = `Bearer ${updatedToken}`;
+    }
+
+    const match = uri.pathname.match(/^storage\/([^/]+)(.*)$/);
+    if (method === 'PUT' && match) {
+      const username = decodeURIComponent(match[1]).split('@')[0];
+      const path = match[2];
+
+      let canWrite = false;
+      if (metadata[HAS_WRITE_CAPACITY_METADATA]) {
+        try {
+          const decoded = symmetricDecrypt(metadata[HAS_WRITE_CAPACITY_METADATA], salt);
+          canWrite = decoded.toString().startsWith(HAS_WRITE_CAPACITY_METADATA_VALUE_PREFIX);
+        } catch {}
+      }
+      if (!canWrite) {
+        getLogger().error(`denied write: user:${username} path:${path} :: insufficient capacity`);
+        this.response.writeHead(507, this._headers);
+        this.response.end();
+        return;
+      }
+    }
+
+    await this._next();
+  };
+
+  getSize (username) {
+    const datapath = this._store.dataPath(username, '/');
+    try {
+      if (!fs.existsSync(datapath)) {
+        return 0;
+      }
+    } catch (err) {
+      return 0;
+    }
+    const size = fastFolderSizeSync(datapath);
```

Comment:
Since this will only work for the filesystem storage back-end, I'm wondering if an extension should also declare which storage back-ends it (currently) supports.

Also, for paid services it would be nice to have per-user limits, of course. But that would be a future feature on top of this one, of course.

Reply:
I'll add a blurb in the README about the current limitation.

Forgot about the coupling, good point.

Yes, my next PR is to bring in payments and IAPs but still very simple payments schedule, just one tier.